modules/imgproc/src/canny.cpp

```diff
@@ -685,14 +717,66 @@ while (borderPeaks.try_pop(m))
     if (!m[mapstep+1])  CANNY_PUSH_SERIAL(m + mapstep + 1);
 }
 
-#else
+// the final pass, form the final image
+const uchar* pmap = map + mapstep + 1;
+uchar* pdst = dst.ptr();
+for (int i = 0; i < src.rows; i++, pmap += mapstep, pdst += dst.step)
+{
+    for (int j = 0; j < src.cols; j++)
+        pdst[j] = (uchar)-(pmap[j] >> 1);
+}
 
+#else
     Mat dx(src.rows, src.cols, CV_16SC(cn));
     Mat dy(src.rows, src.cols, CV_16SC(cn));
 
     Sobel(src, dx, CV_16S, 1, 0, aperture_size, 1, 0, BORDER_REPLICATE);
     Sobel(src, dy, CV_16S, 0, 1, aperture_size, 1, 0, BORDER_REPLICATE);
 
+    CannyImpl(dx, dy, dst, low_thresh, high_thresh, L2gradient);
+#endif
+}
+
+void Canny( InputArray _dx, InputArray _dy, OutputArray _dst,
+                double low_thresh, double high_thresh,
+                bool L2gradient )
+{
+    CV_Assert(_dx.dims() == 2 && _dy.dims() == 2);
```

Comment:
we can check only `_dx.dims() == 2`. For `_dy` the condition will be checked at the line 747